### PR TITLE
[RF] Deprecate RooMinuit class (adapter between RooFit and minuit, got replaced by the more general RooMinimizer)

### PR DESCRIPTION
**Pull request by guitargeek**
See commit descriptions.

### CODE REVIEW

File: roofit/histfactory/inc/RooStats/HistFactory/RooBarlowBeestonLL.h [hunk 1]
```diff
@@ -73,7 +71,7 @@ class RooBarlowBeestonLL : public RooAbsReal {
   // void setAlwaysStartFromMin(Bool_t flag) { _startFromMin = flag ; }
   // Bool_t alwaysStartFromMin() const { return _startFromMin ; }
 
-  //RooMinuit* minuit() { return _minuit ; }
+  //RooMinimizer* minimizer() { return _minimizer ; }
   RooAbsReal& nll() { return const_cast<RooAbsReal&>(_nll.arg()) ; }
   // const RooArgSet& bestFitParams() const ;
   // const RooArgSet& bestFitObs() const ;
```
Comment (hageboeck) [inline roofit/histfactory/inc/RooStats/HistFactory/RooBarlowBeestonLL.h:77]: Maybe clean this up while you are at it ...

Comment (hageboeck) [reply roofit/histfactory/inc/RooStats/HistFactory/RooBarlowBeestonLL.h:77]: It's all code that will just rot because it's never being compiled.

File: roofit/histfactory/inc/RooStats/HistFactory/RooBarlowBeestonLL.h [hunk 2]
```diff
@@ -114,7 +112,7 @@ class RooBarlowBeestonLL : public RooAbsReal {
   TIterator* _oiter ; //! Iterator of profile likelihood output parameter(s)
   */
 
-  // mutable RooMinuit* _minuit ; //! Internal minuit instance
+  // mutable RooMinimizer* _minimizer ; //! Internal minimizer instance
 
   // mutable Bool_t _absMinValid ; // flag if absmin is up-to-date
   // mutable Double_t _absMin ; // absolute minimum of -log(L)
```
Comment (hageboeck) [inline roofit/histfactory/inc/RooStats/HistFactory/RooBarlowBeestonLL.h:118]: ... and here.

File: roofit/histfactory/src/HistoToWorkspaceFactory.cxx [hunk 3]
```diff
@@ -1056,9 +1055,9 @@ namespace HistFactory{
     /*
     RooMsgService::instance().setGlobalKillBelow(RooFit::DEBUG) ;
     profile->getVal();
-    RooMinuit* minuit = ((RooProfileLL*) profile)->minuit();
+    RooMinimizer* minuit = ((RooProfileLL*) profile)->minimizer();
     minuit->setPrintLevel(5) ; // Print MINUIT messages
-    minuit->setVerbose(5) ; // Print RooMinuit messages with parameter 
+    minuit->setVerbose(5) ; // Print RooMinimizer messages with parameter 
                                 // changes (corresponds to the Verbose() option of fitTo()
     */
```
Comment (hageboeck) [inline roofit/histfactory/src/HistoToWorkspaceFactory.cxx:1062]: ... and here.

File: roofit/roofitcore/inc/RooProfileLL.h [hunk 4]
```diff
@@ -62,7 +59,7 @@ class RooProfileLL : public RooAbsReal {
   TIterator* _piter ; //! Iterator over profile likelihood parameters to be minimized 
   TIterator* _oiter ; //! Iterator of profile likelihood output parameter(s)
```
Comment (hageboeck) [inline roofit/roofitcore/inc/RooProfileLL.h:60]: When you see (mutable) iterator members that are not serialised (`//!`), that is almost always a sign that a range-based for loop or a counting loop should be used. Less members, faster, easier to understand.

Just saying, maybe for another day ...

**Review by hageboeck — APPROVED** [review]
LGTM, but I have a few more suggestions.

File: roofit/roofitcore/src/RooProfileLL.cxx [hunk 5]
```diff
@@ -221,15 +184,13 @@ Double_t RooProfileLL::evaluate() const
   _neval = _minimizer->evalCounter() ;
 
   // Restore original values and constant status of observables
-  TIterator* iter = obsSetOrig->createIterator() ;
-  RooRealVar* var ;
-  while((var=dynamic_cast<RooRealVar*>(iter->Next()) ) )  {
-    RooRealVar* target = (RooRealVar*) _obs.find(var->GetName()) ;
+  for(auto const& arg : obsSetOrig) {
+    auto var = dynamic_cast<RooRealVar*>(arg);
+    if(var == nullptr) break;
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooProfileLL.cxx:189]: This pattern was used in a lot of old loops, and although it looks like a double check (both for checking that the loop isn't finished and that the type is correct), it's usually only checking if the loop is finished. In almost all cases, you can `static_cast` and run through the full collection. For good measure, I put an `assert` with a `dynamic_cast` into the loop. It brings extra debug info, and it doesn't cost anything in release builds.

Note also that if they were really checking the types, it wouldn't make any sense if you put `{var, var, otherType, var}` in a collection, because the last var would never be reached. That's most probably a bug.

Long story short: Those loops usually work with range-base and `static_cast`, and they will be much faster this way. Consider this as a standard way to modernise those kinds of loops (and make the `var` `const` whenever you can; here you can't). Also note that although `auto const&` looks like an awesome thing for "normal" collections, it's actually obfuscating the fact that you have a collection of pointers.
```suggestion
  for(auto const arg : obsSetOrig) {
    auto var = static_cast<RooRealVar*>(arg);
    assert(dynamic_cast<RooRealVar*>(arg));
```

Comment (hageboeck) [reply roofit/roofitcore/src/RooProfileLL.cxx:189]: And if you want (I never went there), you can write an adapter that exposes a special kind of iterator:
```c++
for (auto const pvar : CastCollection<const RooRealVar*>(obsSetOrig) ) {
  pvar->GetName();
}
  ```
  Maybe you can get some inspiration from #8421

Comment (guitargeek) [reply roofit/roofitcore/src/RooProfileLL.cxx:189]: Thanks for the comment! Very good point, I was too cautions here and didn't want to change the logic of the code. I'll do it now.

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
@phsft-bot build (ne more round of tests now that `pcepsft10` is fixed)

**Comment by guitargeek** [comment]
Since the ROOT-ubuntu16 tests passed already in the previous round on a machine where it worked, we have now all the tested configurations passing.
